### PR TITLE
Add sys module to `mc-sgx-sync`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**
The sys module contains a `Mutex` implementation for use in SGX enclaves

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #14** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/14" target="_blank">Graphite</a> 
    * **PR #13** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/13" target="_blank">Graphite</a> 
      * **PR #15** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/15" target="_blank">Graphite</a> 
        * **PR #22** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/22" target="_blank">Graphite</a> 
          * **PR #24** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/24" target="_blank">Graphite</a> 
            * **PR #30** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/30" target="_blank">Graphite</a> 
              * **PR #31** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/31" target="_blank">Graphite</a> 
                * **PR #32** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/32" target="_blank">Graphite</a> 
                  * **PR #33** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/33" target="_blank">Graphite</a> 
                    * **PR #34** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/34" target="_blank">Graphite</a> 
                      * **PR #35** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/35" target="_blank">Graphite</a> 
                        * **PR #36** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/36" target="_blank">Graphite</a> 
                          * **PR #39** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/39" target="_blank">Graphite</a>  👈
        * **PR #21** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/21" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/39?utm_source=stack-comment).

File: sync/Cargo.toml [hunk 1]
```diff
@@ -11,5 +11,11 @@ description = "Synchronization primitives for SGX enclaves"
...
 keywords = ["sgx", "no-std"]
 
+[lib]
+# test false due to needing an enclave to fully link
+test = false
+doctest = false
+
 [dependencies]
 mc-sgx-panic = { path = "../panic", version = "=0.1.0" }
+mc-sgx-tstdc = { git = "https://github.com/mobilecoinfoundation/sgx", rev = "12c88f98c6c5b7b71ee7cfe3579422e08eb52cf5" }
```
Comment (nick-mobilecoin) [inline sync/Cargo.toml:21]: This should get changed prior to integration.  It depends on https://github.com/mobilecoinfoundation/sgx/pull/256

File: sync/src/sys/mutex.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/// The mutex backend to use with the common Rust std lib Mutex interface
+pub(crate) struct Mutex {
+    inner: SgxMutex,
+}
+
+unsafe impl Send for Mutex {}
+unsafe impl Sync for Mutex {}
+
+impl Mutex {
```
Comment (nick-mobilecoin) [inline sync/src/sys/mutex.rs:20]: I'm on the fence with this implementation along with https://github.com/mobilecoinfoundation/sgx/pull/256.
Doing it this way keeps the lower level C/SGX API out of this crate, but the abstraction here seems questionable.

The future plan is to take most of https://github.com/rust-lang/rust/blob/master/library/std/src/sync/mutex.rs and have this module contain the SGX specific logic. This means that this module will need to exist it's a matter of deciding if it's this thin abstraction or if the `sgx_thread_mutex_*` calls are invoked directly from this module

**Review by samdealy — APPROVED** [review]

**Review by jcape — APPROVED** [review]
The code here generally seems reasonable, but is the intent to put the poison state into this structure in a later PR?

**Comment by nick-mobilecoin** [comment]
> The code here generally seems reasonable, but is the intent to put the poison state into this structure in a later PR?

Poison state happens in [`std::sync::mutex.rs`](https://github.com/rust-lang/rust/blob/master/library/std/src/sync/mutex.rs), which comes in in #40

File: sync/src/sys/mod.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+//! Platform specific abstraction.
+//!
+//! Provides an abstracted interface over the SGX interface.
+//! This follows the pattern of the
+//! [`std::sys`](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/mod.rs)
+//! module. The difference being this _only_ supports the SGX interface.
+//! Having this allows us to re-use the higher level code modules from
+//! [`::std`](https://github.com/rust-lang/rust/tree/master/library/std/src)
+
+mod mutex;
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply sync/src/sys/mod.rs]: Thanks, I saw that at some point in the past, but forgot about that or got blinded by copy pasta

Module has been updated now

File: sync/Cargo.toml (hunk 1 above)
Comment (nick-mobilecoin) [reply sync/Cargo.toml]: This has been fixed now

**Comment by nick-mobilecoin** [comment]
@varsha888 bumping for re-review due to restack

**Review by varsha888 — APPROVED** [review]

**Comment by nick-mobilecoin** [comment]
⏳ This pull request is set to merge as part of a Graphite merge job
Stack job ID: rnb0zcLhVk2UL31G5ihD.
See details on [graphite.dev](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/39)

**Pull request merged**